packages/react/src/components/Listbox/ListboxOption.tsx

```diff
@@ -0,0 +1,121 @@
+import React, { forwardRef, useEffect, useCallback } from 'react';
+import classnames from 'classnames';
+import { useId } from 'react-id-generator';
+import { useListboxContext } from './ListboxContext';
+import useSharedRef from '../../utils/useSharedRef';
+
+export type ListboxValue = Readonly<string | number | undefined>;
+
+interface ListboxOptionsProps extends React.HTMLAttributes<HTMLElement> {
+  as?: React.ElementType | string;
+  value?: ListboxValue;
+  disabled?: boolean;
+  activeClass?: string;
+}
+
+function isElementPreceding(a: Element, b: Element) {
+  return !!(b.compareDocumentPosition(a) & Node.DOCUMENT_POSITION_PRECEDING);
+}
+
+const ListboxOption = forwardRef<HTMLElement, ListboxOptionsProps>(
+  (
+    {
+      id: propId,
+      className,
+      as: Component = 'li',
+      children,
+      value,
+      disabled,
+      activeClass = 'ListboxOption--active',
+      onClick,
+      ...props
+    },
+    ref
+  ): JSX.Element => {
+    const { active, selected, setOptions, onSelect } = useListboxContext();
+    const listboxOptionRef = useSharedRef<HTMLElement>(ref);
+    const [id] = propId ? [propId] : useId(1, 'listbox-option');
+    const isActive =
+      active !== null && active.element === listboxOptionRef.current;
+    const isSelected =
+      selected !== null && selected.element === listboxOptionRef.current;
+    const optionValue =
+      typeof value !== 'undefined'
+        ? value
+        : listboxOptionRef.current?.innerText;
+
+    useEffect(() => {
+      const element = listboxOptionRef.current;
+
+      setOptions((options) => {
+        const option = { element, value: optionValue };
+        // istanbul ignore next
+        if (!element) return options;
+
+        // Elements are frequently appended, so check to see if the newly rendered
+        // element follows the last element first before any other checks
+        if (
+          !options.length ||
+          isElementPreceding(
+            options[options.length - 1].element,
+            option.element
+          )
+        ) {
+          return [...options, option];
+        }
+
+        for (const opt of options) {
+          if (isElementPreceding(element, opt.element)) {
```

Comment:
It seems like this might not work properly when elements are added. I did some testing and `Node.DOCUMENT_POSITION_PRECEDING` is not the same as "immediately preceding", but this code seems to assume that's how that works.

Reply:
Correct, it's not immediately proceeding.

From testing this was always triggered with a single node at a time (even when mounting/changing multiple nodes). Even when nodes got sorted, the options gets removed from the list and re-added in order so I didn't encounter issues from testing.

Reply:
Even when dynamically adding a new option to the end of the list after the others had already rendered?

Reply:
Yes, that's what lines 57-65 are covering. This is with the assumption that the elements are already in the correct dom order, if the last element in the array precedes the newly inserted element it should get tacked on at the end.